File: packages/jexlang-swift/jexlang-swift/eval/evaluator/JexEvaluator.swift
//
//  JexEvaluator.swift
//  jexlang-swift
//
//  Created by Vinay Kumar on 03/10/25.
//

import Foundation

// MARK: - Evaluator
public class JexEvaluator {

    private var context: [String: JexValue] = [:]
    private var funcsMap: [String: FuncImpl] = [:]
    private var transformMap: [String: TransformImpl] = [:]
    
    private var globalScope: Scope = createGlobalScope()
    
    private lazy var evalVisitor: EvalVisitor = {
        let evaluatorContext = EvaluatorContext(jexEvaluator: self)
        return EvalVisitor(
            scope: globalScope,
            ctx: evaluatorContext,
            funcsMap: self.funcsMap,
            transformsMap: self.transformMap
        )
    }()
    
    private func convertContextToJexValue(context: [String: Any]) -> [String: JexValue] {
        var jexContext: [String: JexValue] = [:]
        if (!context.isEmpty) {
            for (key, value) in context {
                jexContext[key] = JexValueFactory.from(value)
            }
        }
        return jexContext;
    }
    
    private func convertJexValueToAny(jexValueMap: [String: JexValue]) -> [String: Any] {
        var context: [String: Any] = [:]
        for (key, value) in jexValueMap {
            context[key] = value.toObject()
        }
        return context
    }
    
    private func addAllContextValuesIntoGlobalScope(context: [String: JexValue]) throws {
        for (key, value) in context {
            globalScope.declareVariable(key, value: value, isConst: false)
        }
    }
    
    private var cacheParsedTrees: [String: JexLangParser.ProgramContext] = [:]
    private var cacheExpressions: Bool = false
    
    // Holding strong references. If we put this as local variable inside parseExpression. then after completing the function (outside function) self instance is deallocating.
    private var lexer: JexLangLexer!
    private var tokens: CommonTokenStream!
    private var parser: JexLangParser!

    public init(
        context initialContext: [String: Any]? = nil,
        funcsMap initialFuncs: [String: FuncImpl]? = nil,
        transformMap initialTransforms: [String: TransformImpl]? = nil
    ) throws {

        try catchNSException { [weak self] in
            guard let self = self else { return }
            if let initialContext = initialContext {
                self.context = self.convertContextToJexValue(context: initialContext)
            }

            if let initialFuncs = initialFuncs {
                self.funcsMap = initialFuncs
            }

            if let initialTransforms = initialTransforms {
                self.transformMap = initialTransforms
            }

            try self.addAllContextValuesIntoGlobalScope(context: self.context)

            let evaluatorContext = EvaluatorContext(jexEvaluator: self)
            self.evalVisitor = EvalVisitor(
                scope: self.globalScope,
                ctx: evaluatorContext,
                funcsMap: self.funcsMap,
                transformsMap: self.transformMap
            )
        }
    }


    // MARK: - Parse (with caching)
    private func parseExpression(expr: String) throws -> JexLangParser.ProgramContext {
        if cacheExpressions, let cached = cacheParsedTrees[expr] {
            return cached
        }
        
        let input = ANTLRInputStream(expr)
        self.lexer = JexLangLexer(input)
        self.tokens = CommonTokenStream(lexer)
        self.parser = try JexLangParser(tokens)
        let errorListener = JexLangErrorListener()

        // --- Error handling setup ---
        lexer.removeErrorListeners()
        parser.removeErrorListeners()

        lexer.addErrorListener(errorListener)
        parser.addErrorListener(errorListener)
        errorListener.clear()

        // --- Parse ---
        let tree = try parser.program()

        // --- Validate syntax errors ---
        if errorListener.hasErrors() {
            throw errorListener.getSyntaxErrors()[0]
        }

        if cacheExpressions {
            cacheParsedTrees[expr] = tree
        }

        return tree
    }

    // MARK: - Evaluate
    public func evaluate(
        expr: String,
        programScopeVariables: [String: Any]? = nil
    ) throws -> Any {
        return try catchNSException {
            let programContext = try self.parseExpression(expr: expr)
            if let programScopeVariables = programScopeVariables {
                self.evalVisitor.setProgramScopeContext(context: programScopeVariables)
            }
            // Evaluate using the retained parser context
            let value = self.evalVisitor.visit(programContext)

            return value.toObject()
        }
    }
    
    public func setCacheExpressions(_ cacheExpressions: Bool) {
        self.cacheExpressions = cacheExpressions
    }
    
    public func getCacheExpressions() -> Bool {
        return self.cacheExpressions
    }
    
    public func setContextValue(_ key: String, _ value: Any) throws {
        return try catchNSException {
            let jexValue = JexValueFactory.from(value)
            self.context[key] = jexValue
            self.globalScope.assignVariable(key, value: jexValue)
        }
    }
    
    public func setContextValue(_ key: String, _ value: JexValue) throws {
        return try catchNSException {
            self.context[key] = value
            self.globalScope.assignVariable(key, value: value)
        }
    }
    
    @discardableResult
    public func declareContextValue(_ key: String, value: Any?, isConst: Bool = false) throws -> JexValue {
        return try catchNSException {
            let jexValue = JexValueFactory.from(value)
            self.context[key] = jexValue
            self.globalScope.declareVariable(key, value: jexValue, isConst: isConst)
            return jexValue
        }
    }
    
    @discardableResult
    public func declareContextValue(_ key: String, value: JexValue, isConst: Bool = false) throws -> JexValue {
        return try catchNSException {
            self.context[key] = value
            self.globalScope.declareVariable(key, value: value, isConst: isConst)
            return value
        }
    }
    
    @discardableResult
    public func setContextOrDeclareContextValue(
        _ key: String,
        _ value: Any,
        isConst: Bool = false
    ) throws -> JexValue {
        return try catchNSException {
            let jexValue = JexValueFactory.from(value)
            self.context[key] = jexValue
            if (self.globalScope.hasVariable(key)) {
                self.globalScope.assignVariable(key, value: jexValue)
            } else {
                self.globalScope.declareVariable(key, value: jexValue, isConst: isConst)
            }
            return jexValue
        }
    }
    
    public func resetContext() throws {
        return try catchNSException { [weak self] in
            guard let self = self else { return }
            self.context.removeAll()
            self.globalScope = createGlobalScope()
            try self.addAllContextValuesIntoGlobalScope(context: self.context)
            self.evalVisitor = EvalVisitor(
                scope: self.globalScope,
                ctx: EvaluatorContext(jexEvaluator: self),
                funcsMap: self.funcsMap,
                transformsMap: self.transformMap
            )
        }
    }
    
    public func resetFunctions() {
        self.funcsMap.removeAll()
        self.evalVisitor = EvalVisitor(
            scope: self.globalScope,
            ctx: EvaluatorContext(jexEvaluator: self),
            funcsMap: self.funcsMap,
            transformsMap: self.transformMap
        )
    }
    
    public func resetTransforms() {
        self.transformMap.removeAll()
        self.evalVisitor = EvalVisitor(
            scope: self.globalScope,
            ctx: EvaluatorContext(jexEvaluator: self),
            funcsMap: self.funcsMap,
            transformsMap: self.transformMap
        )
    }
    
    public func getContextValue(_ name: String) -> Any? {
        let jexValue = self.context[name] ?? nil
        if let jexValue = jexValue {
            return jexValue.toObject()
        }
        return nil
    }
    
    public func getContextValue(_ name: String, _ asJexValue: Bool) -> JexValue? {
        let jexValue = self.context[name] ?? nil
        if let jexValue = jexValue {
            return jexValue
        }
        return nil
    }
    
    public func getGlobalScopeVariables() -> [String: Any] {
        guard let globalVariables = self.evalVisitor.getGlobalScopeVariables() else {
            return [:]
        }
        return self.convertJexValueToAny(jexValueMap: globalVariables)
    }
    
    public func addFunction(
        name: String,
        function: @escaping FuncImpl
    ) {
        funcsMap[name] = function
        evalVisitor.addFunction(name: name, fn: function)
    }
    
    public func removeFunction(name: String) {
        funcsMap.removeValue(forKey: name)
        evalVisitor.removeFunction(name: name)
    }
    
    public func getAllFunctions() -> [String: FuncImpl]{
        return self.evalVisitor.getAllFunctions()
    }
    
    public func hasFunction(name: String) -> Bool {
        return self.evalVisitor.hasFunction(name)
    }
    
    public func addTransform(
        name: String,
        transform: @escaping TransformImpl
    ) {
        transformMap[name] = transform
        evalVisitor.addTransform(name: name, transform: transform)
    }
    
    public func getAllTransforms() -> [String: TransformImpl] {
        return self.evalVisitor.getAllTransforms()
    }
    
    public func hasTransform(name: String) -> Bool {
        return self.evalVisitor.hasTransform(name)
    }
    
    public func removeTransform(name: String) {
        transformMap.removeValue(forKey: name)
        evalVisitor.removeTransform(name: name)
    }
    
    public func addFunctions(functions: [String: FuncImpl]) {
        for (name, function) in functions {
            self.addFunction(name: name, function: function)
        }
    }
    
    public func addTransforms(transforms: [String: TransformImpl]) {
        for (name, transform) in transforms {
            self.addTransform(name: name, transform: transform)
        }
    }
    
    public func clearCachedParsedExpression() {
        self.cacheParsedTrees.removeAll()
    }
}

File: packages/jexlang-java/src/main/java/io/github/vinaykumar0339/evaluator/JexEvaluator.java
package io.github.vinaykumar0339.evaluator;

import io.github.vinaykumar0339.Utils;
import io.github.vinaykumar0339.context.EvaluatorContext;
import io.github.vinaykumar0339.functions.FuncImpl;
import io.github.vinaykumar0339.grammar.JexLangLexer;
import io.github.vinaykumar0339.grammar.JexLangParser;
import io.github.vinaykumar0339.listeners.JexLangErrorListener;
import io.github.vinaykumar0339.scopes.Scope;
import io.github.vinaykumar0339.transforms.TransformImpl;
import io.github.vinaykumar0339.types.JexValue;
import io.github.vinaykumar0339.visitors.EvalVisitor;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CodePointCharStream;
import org.antlr.v4.runtime.CommonTokenStream;

import java.util.HashMap;
import java.util.Map;

public class JexEvaluator {

    private final Map<String, JexValue> context;
    private final Map<String, FuncImpl> funcsMap;
    private final Map<String, TransformImpl> transformMap;

    private Scope globalScope = Utils.createGlobalScope();
    private EvalVisitor evalVisitor;
    private final JexLangErrorListener errorListener;
    private final Map<String, JexLangParser.ProgramContext> cacheParsedTrees = new HashMap<>();
    private boolean cacheExpressions = false;

    private EvaluatorContext getEvalContext() {
        return new EvaluatorContext(this);
    };

    private Map<String, JexValue> convertContextToJexValue(Map<String, Object> context) {
        Map<String, JexValue> jexContext = new HashMap<>();
        if (context != null) {
            for (Map.Entry<String, Object> entry : context.entrySet()) {
                jexContext.put(entry.getKey(), JexValue.from(entry.getValue()));
            }
        }
        return jexContext;
    }

    private Map<String, Object> convertJexValueToContext(Map<String, JexValue> jexValueMap) {
        Map<String, Object> context = new HashMap<>();
        for  (Map.Entry<String, JexValue> entry : jexValueMap.entrySet()) {
            context.put(entry.getKey(), entry.getValue().toObject());
        }
        return context;
    }

    private void addAllContextValuesIntoGlobalScope(Map<String, JexValue> context) {
        if (context != null) {
            for (Map.Entry<String, JexValue> entry : context.entrySet()) {
                globalScope.declareVariable(entry.getKey(), entry.getValue(), false);
            }
        }
    }

    public JexEvaluator(
            Map<String, Object> context,
            Map<String, FuncImpl> funcsMap,
            Map<String, TransformImpl> transformMap
    ) {
        this.context = context != null ? convertContextToJexValue(context) : new HashMap<>();
        addAllContextValuesIntoGlobalScope(this.context);
        this.funcsMap = funcsMap != null ? funcsMap : new HashMap<>();
        this.transformMap = transformMap != null ? transformMap : new HashMap<>();
        this.evalVisitor = new EvalVisitor(this.globalScope, getEvalContext(), funcsMap, transformMap);
        this.errorListener = new JexLangErrorListener();
    }

    private JexLangParser.ProgramContext parseExpression(String expr) {
        if (this.cacheParsedTrees.containsKey(expr) && this.cacheExpressions) {
            JexLangParser.ProgramContext cachedTree = this.cacheParsedTrees.get(expr);
            if (cachedTree != null) {
                return cachedTree;
            }
        }

        CodePointCharStream charStreams = CharStreams.fromString(expr);
        JexLangLexer lexer = new JexLangLexer(charStreams);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        JexLangParser parser = new JexLangParser(tokens);

        // Remove the default console error listener
        lexer.removeErrorListeners();

        // Add our custom error listener
        this.errorListener.clear();
        lexer.addErrorListener(this.errorListener);

        // Remove default console error listener from parser too
        parser.removeErrorListeners();
        parser.addErrorListener(this.errorListener);

        // Parse the expression
        JexLangParser.ProgramContext programContext = parser.program();

        // Check if we have any syntax errors
        if (this.errorListener.hasErrors()) {
            throw this.errorListener.getSyntaxErrors().get(0);
        }

        if (this.cacheExpressions) {
            this.cacheParsedTrees.put(expr, programContext);
        }

        return programContext;
    }

    public void setCacheExpressions(boolean cacheExpressions) {
        this.cacheExpressions = cacheExpressions;
    }

    public boolean getCacheExpressions(boolean cacheExpressions) {
        return this.cacheExpressions;
    }

    public Object evaluate(String expr) {
        return evaluate(expr, Map.ofEntries());
    }

    public Object evaluate(
            String expr,
            Map<String, Object> programScopeVariables
    ) {
        JexLangParser.ProgramContext programContext = parseExpression(expr);
        evalVisitor.setProgramScopeContext(programScopeVariables);
        JexValue value = evalVisitor.visit(programContext);
        if (value != null) {
            return value.toObject();
        }
        return null;
    }

    public void setContextValue(String name, Object value) {
        JexValue jexValue = JexValue.from(value);
        this.context.put(name, jexValue);
        this.globalScope.assignVariable(name, jexValue);
    }

    public JexValue declareContextValue(String name, Object value, boolean isConst) {
        JexValue jexValue = JexValue.from(value);
        this.context.put(name, jexValue);
        this.globalScope.declareVariable(name, jexValue, isConst);
        return jexValue;
    }

    public JexValue setContextOrDeclareContextValue(String name, Object value, boolean isConst) {
        JexValue jexValue = JexValue.from(value);
        this.context.put(name, jexValue);
        if (this.globalScope.hasVariable(name)) {
            this.globalScope.assignVariable(name, jexValue);
        } else {
            this.globalScope.declareVariable(name, jexValue, isConst);
        }
        return jexValue;
    }

    public void resetContext() {
        this.context.clear();
        this.globalScope = Utils.createGlobalScope();
        addAllContextValuesIntoGlobalScope(this.context);
        this.evalVisitor = new EvalVisitor(this.globalScope, getEvalContext(), this.funcsMap, this.transformMap);
    }

    public void resetFunctions() {
        this.funcsMap.clear();
        this.evalVisitor = new EvalVisitor(this.globalScope, getEvalContext(), this.funcsMap, this.transformMap);
    }

    public void resetTransforms() {
        this.transformMap.clear();
        this.evalVisitor = new EvalVisitor(this.globalScope, getEvalContext(), this.funcsMap, this.transformMap);
    }

    public Object getContextValue(String name) {
        JexValue value = this.context.get(name);
        if (value != null) {
            return value.toObject();
        }
        return null;
    }

    public Map<String, Object> getGlobalScopeVariables() {
        Map<String, JexValue> globalVariables = this.evalVisitor.getGlobalScopeVariables();
        if (globalVariables != null) {
            return this.convertJexValueToContext(globalVariables);
        }
        return Map.of();
    }

    public void addFunction(String name, FuncImpl function) {
        funcsMap.put(name, function);
        evalVisitor.addFunction(name, function);
    }

    public Map<String, FuncImpl> getAllFunctions() {
        return this.evalVisitor.getAllFunctions();
    }

    public boolean hasFunction(String name) {
        return this.evalVisitor.hasFunction(name);
    }

    public Map<String, TransformImpl> getAllTransforms() {
        return this.evalVisitor.getAllTransforms();
    }

    public boolean hasTransform(String name) {
        return this.evalVisitor.hasTransform(name);
    }

    public void addTransform(String name, TransformImpl transform) {
        transformMap.put(name, transform);
        evalVisitor.addTransform(name, transform);
    }

    public void addFunctions(Map<String, FuncImpl> functions) {
        if (functions != null) {
            for (Map.Entry<String, FuncImpl> entry : functions.entrySet()) {
                addFunction(entry.getKey(), entry.getValue());
            }
        }
    }

    public void addTransforms(Map<String, TransformImpl> transforms) {
        if (transforms != null) {
            for (Map.Entry<String, TransformImpl> entry : transforms.entrySet()) {
                addTransform(entry.getKey(), entry.getValue());
            }
        }
    }

    public void clearCachedParsedExpressions() {
        cacheParsedTrees.clear();
    }
}
